Rules/UseUsingScopeModifierInNewRunspaces.cs

```diff
@@ -0,0 +1,430 @@
+        /// </returns>
+        public override AstVisitAction VisitScriptBlockExpression(ScriptBlockExpressionAst scriptBlockExpressionAst)
+        {
+            if (!(scriptBlockExpressionAst.Parent is CommandAst commandAst))
+            {
+                return AstVisitAction.Continue;
+            }
+
+            var cmdName = commandAst.GetCommandName();
+            if (cmdName == null)
+            {
+                // Skip for situations where command name cannot be resolved like `& $commandName -ComputerName -ScriptBlock { $foo }` 
+                return AstVisitAction.Continue;
+            }
+            
+            var scriptBlockParameterAst = commandAst.CommandElements[
+                    commandAst.CommandElements.IndexOf(scriptBlockExpressionAst) - 1] as CommandParameterAst;
+
+            if (IsInlineScriptBlock(cmdName) || 
+                IsJobScriptBlock(cmdName, scriptBlockParameterAst) ||
+                IsForeachScriptBlock(cmdName, scriptBlockParameterAst) ||
+                IsInvokeCommandComputerScriptBlock(cmdName, commandAst) ||
+                IsDSCScriptResource(cmdName, commandAst))
+            {
+                AnalyzeScriptBlock(scriptBlockExpressionAst);
+                return AstVisitAction.SkipChildren;
+            }
+
+            if (IsInvokeCommandSessionScriptBlock(cmdName, commandAst))
+            {
+                var sessionName = GetSessionName(commandAst);
+
+                var varsInLocalAssignments = FindVarsInAssignmentAsts(scriptBlockExpressionAst);
+                if (varsInLocalAssignments != null)
+                {
+                    AddAssignedVarsToSession(sessionName, varsInLocalAssignments);
+                }
+
+                GenerateDiagnosticRecords(
+                    FindNonAssignedNonUsingVarAsts(
+                        scriptBlockExpressionAst, 
+                        GetAssignedVarsInSession(sessionName)));
+
+                return AstVisitAction.SkipChildren;
+            }
+            
+            return AstVisitAction.Continue;
+        }
+
+        /// <summary>
+        /// FindVarsInAssignmentAsts: Retrieves all assigned variables from an Ast:
+        /// Example: `$foo = "foo"` ==> the VariableExpressionAst for $foo is returned
+        /// </summary>
+        /// <param name="ast"></param>
+        private static IEnumerable<VariableExpressionAst> FindVarsInAssignmentAsts(Ast ast)
+        {
+            // Find all variables that are assigned within this ast
+            return ast.FindAll(
+                    predicate: a => a is VariableExpressionAst varExpr &&
+                                    varExpr.Parent is AssignmentStatementAst assignment &&
+                                    assignment
+                                        .Left
+                                        .Equals(varExpr),

```

Comment:
This can be bypassed with something like:

```
[int]$x = 10
```

The assignment is valid, but the LHS of the assignment doesn't equal the variable AST

Reply:
Great catch! You are absolutely right. I changed two tests to make sure we keep checking for this. I'll have to take this into account when refactoring to visitor method.